File: HNOI_2003_激光炸弹.py
n, r = map(int, input().split())
r = min(5001, r)
ans = 0
all = [[0 for _ in range(5005)] for _ in range(5005)]
for _ in range(n):
    x, y, v = map(int, input().split())
    all[x + 1][y + 1] += v
for i in range(1, 5001):
    for j in range(1, 5001):
        all[i][j] += all[i - 1][j] + all[i][j - 1] - all[i - 1][j - 1]
        if i >= r and j >= r:
            ans = max(ans, all[i][j] - all[i - r][j] - all[i][j - r] + all[i - r][j - r])
print(ans)

File: HNOI_2003_激光炸弹.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl '\n'
#define int long long
signed main() {
    int n, r;
    cin >> n >> r;
    r = min(5001ll, r);
    int ans = 0;
    int all[5005][5005];
    for (int i = 0; i < n; ++i) {
        int x, y, v;
        cin >> x >> y >> v;
        all[++x][++y] += v;
    }
    for (int i = 1; i <= 5001; ++i) {
        for (int j = 1; j <= 5001; ++j) {
            all[i][j] += all[i - 1][j] + all[i][j - 1] - all[i - 1][j - 1];
            if (i >= r and j >= r) {
                ans = max(ans, all[i][j] - all[i - r][j] - all[i][j - r] + all[i - r][j - r]);
            }
        }
    }
    cout << ans;
    return 0;
}
